semver.md

```diff
@@ -59,36 +62,33 @@ interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
 could be declared in the code itself or exist strictly in documentation.
 However it is done, it SHOULD be precise and comprehensive.
 
-1. A normal version number MUST take the form X.Y.Z where X, Y, and Z are
-non-negative integers, and MUST NOT contain leading zeroes. X is the
-major version, Y is the minor version, and Z is the patch version.
+1. A normal version number MUST take the form 0.X.Y, 0.0.X, or X.Y.Z where
+X, Y, and Z are non-negative integers, and MUST NOT contain leading zeroes.
+X is the major version, Y is the minor version, and Z is the patch version.
 Each element MUST increase numerically. For instance: 1.9.0 -> 1.10.0 -> 1.11.0.
 
 1. Once a versioned package has been released, the contents of that version
 MUST NOT be modified. Any modifications MUST be released as a new version.
 
-1. Major version zero (0.y.z) is for initial development. Anything MAY change
-at any time. The public API SHOULD NOT be considered stable.
-
-1. Version 1.0.0 defines the public API. The way in which the version number
-is incremented after this release is dependent on this public API and how it
-changes.
```

Comment:
I think the other changes are OK but the removal of these sections would, in my view, be a serious error.

Reply:
That's basically the entire point. The distinction of "v1.0.0" isn't valuable or meaningful, and in practice having it has been harmful.

Reply:
It isn't meaningful to you, but I'd argue that there are plenty of people to whom it is a meaningful distinction. To me it's the keystone to the philosophy of semver. The spec hints at this heavily when in the FAQs:

> If even the tiniest backward incompatible changes to the public API require a major version bump, won’t I end up at version 42.0.0 very rapidly?

The way I interpret the answer is "if there are so many APIs in your package that you're realistically in danger of ending up at v42, you really should have split apart multiple packages with more focused responsibilities".

I think that by making 0.42.0 the same thing as 42.0.0 you're removing the key force the agreement exerts that tends to create strong well-factored ecosystems of code, which in my mind was the largest part of the purpose of this document

Reply:
It's just a number, it doesn't matter if you have v42 or v153. Breaking changes should be avoided in every case regardless, but that's an opinion that doesn't belong in this document. (and that doesn't match my understanding of its purpose, which is to safely convey breakage, not to inhibit it)

Reply:
But v153 does inhibit breakage. It usually means that there are several unrelated APIs inside, e.g. all the helper methods from a utility library, or all the components in a whole design system. Most of the time a bump from v152 to v153 means that 49 of the components in a design system had no breaking changes and one of them did. This is not a situation where semver has failed to protect the publisher, but it is a situation where the it has failed to protect the user, who got 49 false positive reports of changes alongside one real change. This is no longer a valuable tool for helping users understand what has *not* changed about their own software.

Reply:
To be clear, I totally agree with your philosophy here - a major bump when a breaking change doesn't affect you is harmful.

This happens constantly in a way that's entirely unrelated to "is it v1 or not", which is why I think this concern is unrelated to that philosophy. The semver spec simply doesn't address nor attempt to address that concern (nor should it, I suspect)

Reply:
If v3 of the spec is breaking, why not just outlaw versions starting with 0? Wouldn't that solve everyone's problems to their satisfaction? It would be clear that any triplet starting with `0` is not semver v3.

The only problem with it is that it isn't a change that can be retroactively applied without opt-in from the users, but I see that as ideal. The great promise of semver is that it allows you to make breaking changes because people who don't want to be broken have the chance to opt out.

Reply:
The benefit is that it requires no changes at all. Semver already requires versions to start with `1`.

Reply:
No, it doesn’t require that.

Outlawing versions that already exist would be incredibly harmful; that should be a nonstarter.

Reply:
I'm saying the spec exactly as written is already correct. These version are not "outlawed", the `breaking.changing.fixing` triplet pattern simply doesn't apply to them. To get that protection you have to opt in by making at least one official release.

Reply:
The fact that semver explicitly doesn't say what semantics apply to any version starting with 0.x is exactly the freedom you need to be able to solve your problem: defining a meaning for those 0.x version numbers that is consistent with the npm ecosystem's historical expectations. The only problem I have here at all is the idea that the two things are one document: that your changes overwrite the underlying spec instead of coexisting with it as they easily could.